### PR TITLE
refCounted: unittest now also checks object destruction.

### PR DESCRIPTION
**Pull request by DoctorNoobingstoneIPresume**
In the `unittest`, there was a comment:

```
     // file gets properly closed when last reference is dropped 
```

The `unittest` has now been modified so that the statement in the comment is actually tested.

For this, I have moved some of the lines inside a new scope (i.e. a new pair of braces). I have not re-indented the code in order for the Git history to be preserved (e.g. output of `git blame`, `git diff`, `git log -p`).

### CODE REVIEW

File: std/typecons.d [hunk 1]
```diff
@@ -6850,24 +6850,32 @@ RefCounted!(T, RefCountedAutoInitialize.no) refCounted(T)(T val)
...
 
+    assert(File.nDestroyed == 0);
     auto file = File("name");
     assert(file.name == "name");
     // file cannot be copied and has unique ownership
     static assert(!__traits(compiles, {auto file2 = file;}));
+    assert(File.nDestroyed == 0);
 
     // make the file refcounted to share ownership
+    {
```
Comment (RazvanN7) [inline std/typecons.d:6867]: I agree that git history is important, however, it does not represent a reason to violate existing guidelines for code identation. Please insert a tab for this code.

File: std/typecons.d [hunk 2]
```diff
@@ -6850,24 +6850,32 @@ RefCounted!(T, RefCountedAutoInitialize.no) refCounted(T)(T val)
...
     static struct File
     {
+        static size_t nDestroyed;
         string name;
         @disable this(this); // not copyable
-        ~this() { name = null; }
+        ~this() { name = null; ++nDestroyed; }
     }
 
+    assert(File.nDestroyed == 0);
```
Comment (RazvanN7) [inline std/typecons.d:6859]: This line is not necessary.

File: std/typecons.d [hunk 3]
```diff
@@ -6850,24 +6850,32 @@ RefCounted!(T, RefCountedAutoInitialize.no) refCounted(T)(T val)
...
     auto file = File("name");
     assert(file.name == "name");
     // file cannot be copied and has unique ownership
     static assert(!__traits(compiles, {auto file2 = file;}));
+    assert(File.nDestroyed == 0);
```
Comment (RazvanN7) [inline std/typecons.d:6864]: Neither is this one.

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
I'm not sure this addition is necessary. The purpose of the comment was to inform the unittest reader that it is not necessary to close the file manually, not that  destruction should be checked.

Also, adding the `ndestroyed` field is kind of redundant. Isn't the `rcFile.refCountedStore.refCount` field specifically designed for that?
Checking for `refCount` and `nDestroyed` is redundant in this case.

**Comment by DoctorNoobingstoneIPresume** [comment]
> I'm not sure this addition is necessary. The purpose of the comment was to inform the unittest reader that it is not necessary to close the file manually, not that destruction should be checked.

The user is informed that the reason why it is not necessary to close the file manually is that the file has been automatically closed.

This test ensures that the file has indeed been automatically closed (and therefore that the user has not been misled into believing manual closing is not needed).

You are correct: the user does not need to check destruction of his/her objects. The test does not encourage the user to do this kind of instrumentation-and-checking in his/her code. Instead, the test is just a black-box check that the `RefCounted` implementation is correct -- with regard to automatic deletion of the payload object too.

> 
> Also, adding the `ndestroyed` field is kind of redundant. Isn't the `rcFile.refCountedStore.refCount` field specifically designed for that?
> Checking for `refCount` and `nDestroyed` is redundant in this case.

I believe we cannot access the `rcFile.refCountedStore.refCount` field after `rcFile` has been deleted.

Also, using `nDestroyed` provides a sort of black-box testing (without relying on the current `RefCounted` implementation, e.g. field names).

File: std/typecons.d (hunk 2 above)
Comment (DoctorNoobingstoneIPresume) [reply std/typecons.d]: I have now removed one of the tests for zero...

File: std/typecons.d (hunk 3 above)
Comment (DoctorNoobingstoneIPresume) [reply std/typecons.d:6864]: ... but I have kept the other comparison against zero, as I feel it helps clarify the behaviour -- as per newly-added comments too.

I hope it is okay with you.  (-:

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**